shared/src/sources/balancer/graph_api.rs

```diff
@@ -75,13 +78,42 @@ impl BalancerSubgraphClient {
                 pools_by_factory
                     .entry(pool.factory.unwrap_or_default())
                     .or_default()
-                    .push(pool.into_registered(block_number)?);
+                    .push(pool.into_registered_pool(block_number)?);
             }
 
             has_next_page
         } {}
 
-        Ok(RegisteredWeightedPools {
+        let mut last_id = H256::default();
+        while {
+            let page = self
+                .0
+                .query::<pools_query::Data<StableToken>>(
+                    pools_query::STABLE_POOL_QUERY,
+                    Some(json_map! {
+                        "block" => block_number,
+                        "pageSize" => QUERY_PAGE_SIZE,
+                        "lastId" => json!(last_id),
+                    }),
+                )
+                .await?
+                .pools;
+            let has_next_page = page.len() == QUERY_PAGE_SIZE;
+            if let Some(last_pool) = page.last() {
+                last_id = last_pool.id;
+            }
+
+            for pool in page {
+                pools_by_factory
+                    .entry(pool.factory.unwrap_or_default())
+                    .or_default()
+                    .push(pool.into_registered_pool(block_number)?);
```

Comment:
Should this use the `?` operator? That means the whole function fails if into_registered_pool fails which in turn only fails if `anyhow!("unsupported token with more than 18 decimals")`. I think it would be better to only skip this one pool in that case and completely abort the function. In my refactor suggestion above I left a `todo!` for this case. If we wanted to just skip this pool then it would become a `continue`.

Reply:
Realistically I think that expect would probably work best here (except for the case you pointed out - with the decimals). However, it is also unlikely for this to happen since we are always operating on known (deployed) balancer pools and they do not support tokens that would cause this error. So I would say we are safe to `expect` here.

Reply:
In fact, I should really change this into an expect as well.

```rust
18u8.checked_sub(token.decimals).ok_or_else(|| anyhow!("unsupported token with more than 18 decimals"))
```

Reply:
Usually I am hesitant to have an unwrap / expect anywhere that handles third party data (any data that isn't produced from our code) because we don't want to allow a bug in someone else's code to cause a panic in ours.
In this case if it is obvious from the smart contracts that this can't happen then it is probably ok but if it's not too much work I'd still slightly prefer to bubble up the error or error log it.

Reply:
This isn't exactly "third party" since it is coming from a smart contract and those contracts do not allow pool creation for  tokens with more than 18 decimals.

Reply:
There could still be a bug in their smart contract code or even a bug in the ethereum node so that it returns a bogus value there. For example when we are parsing json from an external server even if their documentation says that the response always has an "id" field  and we can verify that in their  code that this is true I still wouldn't like to `expect` this.
More realistic in this case I could see that their contracts are upgradable in which case an upgrade could break the invariant.

Reply:
https://github.com/balancer-labs/balancer-v2-monorepo/blob/0328ed575c1b36fb0ad61ab8ce848083543070b9/pkg/pool-utils/contracts/BasePool.sol#L476-L487

Reply:
Ok then... I will change it back...

Reply:
I replaced your todo with `tracing::error`, then `continue` and some comment explaining the situation with a link.